### PR TITLE
Major refactor

### PR DESCRIPTION
**Pull request by makara**
Testing. Not there yet.

### CODE REVIEW

**Comment by makara** [comment]
@xavierchow Good to go for now. Please have a look. For next steps I would do:

- Better connection handling
- More pluggable "strategies"
- Higher level concepts like "Topic"

File: .gitignore [hunk 1]
```diff
@@ -5,6 +5,7 @@
 !.travis.yml
 
 *.orig
+*.lock
```
Comment (xavierchow) [inline .gitignore]: Please exclude yarn.lock

File: lib/Producer.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+  }
+
+  connect(callback) {
+    this.conns = [];
+    this._closed = false;
+    if (this.lookupd) {
+      return nsqlookup(this.lookupd).then((nodes) => {
+        return Promise.map(nodes, (node) => {
+          return this.connectNsqd(node.broadcast_address, node.tcp_port, this.opts);
+        }).then((error) => {
```
Comment (xavierchow) [inline lib/Producer.js]: The argument `error` here is not needed.

File: lib/Producer.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    if (options.retry && this.strategy === Producer.FAN_OUT) {
+      return Promise.reject(new Error('Retry on produce level is not supported in fanout strategy'))
+        .asCallback(callback);
+    }
+
+    if (options.retry) {
+      return Promise.resolve(promiseRetry({
+        retries: options.retry === true ? 10 : options.retry
```
Comment (xavierchow) [inline lib/Producer.js]: The logic is not same as before, here the whole option `options.retry` would be set as `retries`, which means callers lost the capability of setting other fields(factor, maxTimeout, etc.)

File: lib/api/Nsqlookupd.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+'use strict'
+
+const Base = require('./Base');
+
+module.exports = class Nsqlookupd extends Base {
```
Comment (xavierchow) [inline lib/api/Nsqlookupd.js]: This class is only used for test?

File: lib/api/Nsqd.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+'use strict'
+
+const Base = require('./Base');
+
+module.exports = class Nsqd extends Base {
```
Comment (xavierchow) [inline lib/api/Nsqd.js]: This class is only used for test?

**Comment by xavierchow** [comment]
I think we should keep yarn.lock to provide an alternative for yarn users.

**Comment by makara** [comment]
Fixed things other than the API libs, which are exposed, and could be used for whatever like for example #10 at some point.

**Comment by xavierchow** [comment]
LGTM

**Pull request merged**